File: 4_group_anagram/group_anagram.java
import java.util.*;

public class GroupAnagrams {
    
    public static List<List<String>> groupAnagrams(String[] strs) {
        // Map where key is the sorted word and value is the list of anagrams
        HashMap<String, List<String>> anagramMap = new HashMap<>();

        for (String word : strs) {
            // Convert the word to a char array and sort it
            char[] charArray = word.toCharArray();
            Arrays.sort(charArray);
            String sortedWord = new String(charArray);

            // If the sorted word is not already a key, add a new list
            if (!anagramMap.containsKey(sortedWord)) {
                anagramMap.put(sortedWord, new ArrayList<>());
            }

            // Add the original word to the corresponding list
            anagramMap.get(sortedWord).add(word);
        }

        // Return the grouped values as a list of lists
        return new ArrayList<>(anagramMap.values());
    }

    // Example usage
    public static void main(String[] args) {
        String[] input = {"act", "pots", "tops", "cat", "stop", "hat"};
        List<List<String>> grouped = groupAnagrams(input);
        System.out.println(grouped);
    }
}

File: 4_group_anagram/group_anagram.py
from typing import List
from collections import defaultdict
# group anagram leetcode question 
def groupAnagrams(strs: List[str]) -> List[List[str]]:
    anagram_map = defaultdict(list)  # default value is an empty list

    for word in strs:
        # Sort word to find its "signature"
        sorted_word = ''.join(sorted(word))
        # Group the word using the sorted form
        anagram_map[sorted_word].append(word)

    # Return only the grouped values
    return list(anagram_map.values())

# Example
print(groupAnagrams(["act", "pots", "tops", "cat", "stop", "hat"]))
